Break out config file loading into separate script

ddohler opened the pull request:
## Overview

This should make deployment easier by allowing data loading to happen
without dropping the database and running migrations.

### Checklist

- [x] PR has a name that won't get you publicly shamed for vagueness
- [ ] ~Files changed in the PR have been `yapf`-ed for style violations~

### Notes

I think this is what we discussed but I'm assigning extra reviewers to make sure this works for everyone.

## Testing Instructions

 * (Probably) comment out `clear_database` from `scripts/configure` if you already have data
 * `vagrant ssh` and run both `scripts/configure` and `scripts/load_configured_data`, and make sure they both work as expected.

@tnation14 isn't on the repo yet but I invited him and will add him as a reviewer when he confirms.

ddohler committed:
Break out config file loading into separate script

This should make deployment easier by allowing data loading to happen
without dropping the database and running migrations.

kshepard reviewed (changes requested):
This looks pretty close to what I was expecting. `load_configured_data` looks great. I wonder if `configure` should be split out a little more though. Just thinking from the perspective of someone who wants to deploy a non-VA instance. I think the deployment script will want to run the `clear_database` function, though not the rest of `configure`. And on the user's development environment, they will want to do something similar -- they will be manually assembling/downloading the data zip file, and would want to run the `clear_database` function, followed by `load_configured_data`. So maybe the `clear_database` function just needs to be split out into it's own script?

pcaisse commented on scripts/configure:
```diff
@@ -23,14 +24,11 @@ function load_dev_data() {
-        exec -T django \
-        unzip -o /data/VA_data.zip -d /data
-
+function clear_database() {
```
I think I like `recreate_database` better to make it clear the database will be dropped /nitpick

pcaisse reviewed (commented):
I see there as being 4 steps you might want or need to do separately:

1. Recreate (drop and create) the database
1. Run migrations
1. Fetch and load data
1. Configure

I don't really like the "scripts to rule them all" pattern because each step is a bash script which adds a lot of boilerplate (eg. `set -e` at the top of each and every file) and they're not easily composable. The idea of a this nice mapping of "things you always want to do on a project" (eg. boostrap, setup, update) and a bash script breaks down pretty quickly as we're seeing. It would be nice to have granular control over what steps happen and how to group them beyond writing a specific script that groups all of those steps or running multiple scripts.

tnation14 commented on scripts/configure:
```diff
@@ -23,14 +24,11 @@ function load_dev_data() {
         wget -q \
-            -O /data/VA_data.zip \
+            -O /data/districtbuilder_data.zip \
             https://s3.amazonaws.com/districtbuilderdata/VA_data.zip
```
I think we might want to parametrize the `DISTRICTBUILDER_DATA_URL` as well. Additionally, did we agree to assume that the data would be stored on S3? If so, I'd rather use `aws s3 cp` over `wget` so that the data bucket doesn't have to be publicly accessible.

ddohler replied:
My recollection is that we ended up agreeing that the data would be placed in a designated local directory, which is why I opted not to parametrize this, because I think this script is unique to development setup. I think the same thing would apply to the `wget` vs. `s3` choice, but if I'm misremembering, let me know.

tnation14 replied:
I don't think you're misremembering, that sounds right to me.

tnation14 reviewed (commented):
Looks good!

> The idea of a this nice mapping of "things you always want to do on a project" (eg. boostrap, setup, update) and a bash script breaks down pretty quickly as we're seeing

I think that although STRTA vary in implementation across projects, the high level idea of needing to perform certain tasks to setup your dev environment has held true for the most part. This particular project is causing some complexity because we're trying to use the same scripts to setup Dev and Prod environments, which is generally not the case. That said, it may be worth it to move this functionality into separate deployment scripts, and just let the dev STRTA remain as they previously were.

ddohler commented:
I think @tnation14 's thought about avoiding mixing STRTA and deployment is a good one, so I'm going to refactor to add these as scripts under `/deployment` as well as `/scripts`. I'll leave things split out on the `scripts` side as well to provide ourselves with a starting point for developing and deploying multiple instances. That will result in some duplication between the deployment version of `load_configured_data` and the STRTA version, but I think that's fine because they'll probably evolve away from each other as we develop our deployment processes and this will give them the flexibility to grow independently.

ddohler commented:
Okay, I've updated this. Looking at the development scripts issue, I noticed that we're not really following what STRTA says that our scripts should be doing, which I think led to some of the database work being done in the "wrong" place. I've refactored with the following changes:
- `load_configured_data` has been copied to `deployment/scripts/load_configured_data.sh`
- `scripts/setup` has been renamed to `scripts/bootstrap`
- `scripts/update` has been renamed to `scripts/setup` and now includes running migrations after containers are built.
- `scripts/configure` has been renamed `scripts/configure_va_demo`
- README and `.travis.yml` have been updated appropriately

This should address all the use cases we're considering, I think:
- In deployment, we can place the shapefile in a directory and run `deployment/scripts/load_configured_data.sh` on it. If we need to make changes to this script for deployment, we can, because it's a separate script.
- In local development, setting up the demo VA data is basically the same as before:
    1. `scripts/bootstrap`
    2. `scripts/setup`
    3. `scripts/configure_va_demo`
- But if you want to do development with a different state and config file, then you can:
    1. `scripts/bootstrap`
    2. `scripts/setup` <-- Leaves you with a migrated and ready database
    3. <Manually set up your config.xml and shapefile>
    4. `scripts/load_configured_data`

One thing that I thought about doing, but didn't, is having `scripts/setup` wipe the database before running migrations, and adding a separate `scripts/update` that just runs migrations. I think this could become useful if we find ourselves frequently switching between development on different DistrictBuilder deployments. However, we're not quite there yet, and adding functionality that could create 12-hour re-configuration headaches for anyone who accidentally runs it didn't seem like a step forward. If we decide we really need that functionality then we can add a warning to the script, but for the moment it didn't seem like it would be very useful.

This is ready for another look, I think.

tnation14 commented on .travis.yml:
```diff
@@ -9,7 +9,7 @@ install:
   - sed -i 's/MAP_SERVER_ADMIN_PASSWORD=/MAP_SERVER_ADMIN_PASSWORD=geoserver/' .env
   - cp django/publicmapping/config/config.dist.xml django/publicmapping/config/config.xml
-  - ./scripts/update
+  - ./scripts/setup
```
Minor change, but according to https://github.com/azavea/architecture/blob/master/doc/arch/adr-0000-scripts-to-rule-them-all.md, I think we want the script that provisions the VM to be `scripts/setup`.The script that builds containers/runs migrations can stay named `scripts/update`.

kshepard replied:
For your errors above, I wonder if you need to do the couple steps mentioned in the README:
```
$ cp django/publicmapping/config/config.dist.xml django/publicmapping/config/config.xml
$ # Copy .env file and add password/modify as necessary
$ # NOTE: Leave MAP_SERVER_ADMIN_PASSWORD the same for now
$ cp django/publicmapping/.env.sample django/publicmapping/.env
```

pcaisse replied:
I wonder if it might be helpful to have a `Makefile` that we could use where the targets would have dependencies on these configuration files that need to be in place to run the app.

Alternatively, we could make the `settings.py` file a bit smarter by catching exceptions and outputting more helpful messaging when `docker-compose up` fails (eg. "Did you forget to do a build?").

ddohler replied:
@tnation14 good call; I had forgotten we had our own STRTA ADR and was going off the GitHub spec, which includes a `bootstrap` script, whereas ours doesn't. I've shifted everything back to use `setup` and `update`.

ddohler replied:
@pcaisse I've gone ahead and added a try/except to settings which prints a more helpful message if `config_settings.py` isn't found.

tnation14 replied:
Turns out that `config_settings.py` was getting overridden by `docker-compose` volume mounts, so Django couldn't find the file during initialization. I was able to get the script to succeed from a clean environment by adding this command to `scripts/update`:

```diff
diff --git a/scripts/update b/scripts/update
index 6301c3bc..bda76076 100755
--- a/scripts/update
+++ b/scripts/update
@@ -31,6 +31,14 @@ then
         usage
     else
         build_containers
+
+        # Write config_settings to the filesystem so it doesn't get
+        # overwritten by docker-compose volume mounts
+        docker-compose run --rm --entrypoint "bash -c" django \
+            "python -m district_builder_config.generate_settings \
+                config/config.xsd \
+                config/config.xml \
+                publicmapping/config_settings.py"
         docker-compose up -d postgres django
         run_migrations
     fi
```

kshepard replied:
Cool, it seems this is the same issue discussed here: https://github.com/PublicMapping/DistrictBuilder/pull/512#pullrequestreview-96657452

That's great that you likely came up with a fix.

ddohler replied:
I'm going to add this into this PR since the issue seems to be afflicting Travis.

tnation14 commented:
Not sure if this within the scope of the PR or just a function of setting up a new dev environment, but I can't seem to get the `django` container to start:

```bash
$ vagrant ssh
vagrant@vagrant-trusty-64:/vagrant $ ./scripts/setup
...
Running migrations
Traceback (most recent call last):
  File "./manage.py", line 21, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 308, in execute
    settings.INSTALLED_APPS
  File "/usr/local/lib/python2.7/site-packages/django/conf/__init__.py", line 56, in __getattr__
    self._setup(name)
  File "/usr/local/lib/python2.7/site-packages/django/conf/__init__.py", line 41, in _setup
    self._wrapped = Settings(settings_module)
  File "/usr/local/lib/python2.7/site-packages/django/conf/__init__.py", line 110, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
  File "/usr/local/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/usr/src/app/publicmapping/settings.py", line 235, in <module>
    from publicmapping.config_settings import *
ImportError: No module named config_settings
```

Even after rebuilding the container, `config_settings.py` isn't there, but the container throws a different error:

```bash
django_1     | Traceback (most recent call last):
django_1     |   File "/usr/local/lib/python2.7/site-packages/gunicorn/arbiter.py", line 578, in spawn_worker
django_1     |     worker.init_process()
django_1     |   File "/usr/local/lib/python2.7/site-packages/gunicorn/workers/ggevent.py", line 190, in init_process
django_1     |     super(GeventWorker, self).init_process()
django_1     |   File "/usr/local/lib/python2.7/site-packages/gunicorn/workers/base.py", line 126, in init_process
django_1     |     self.load_wsgi()
django_1     |   File "/usr/local/lib/python2.7/site-packages/gunicorn/workers/base.py", line 135, in load_wsgi
django_1     |     self.wsgi = self.app.wsgi()
django_1     |   File "/usr/local/lib/python2.7/site-packages/gunicorn/app/base.py", line 67, in wsgi
django_1     |     self.callable = self.load()
django_1     |   File "/usr/local/lib/python2.7/site-packages/gunicorn/app/wsgiapp.py", line 65, in load
django_1     |     return self.load_wsgiapp()
django_1     |   File "/usr/local/lib/python2.7/site-packages/gunicorn/app/wsgiapp.py", line 52, in load_wsgiapp
django_1     |     return util.import_app(self.app_uri)
django_1     |   File "/usr/local/lib/python2.7/site-packages/gunicorn/util.py", line 352, in import_app
django_1     |     __import__(module)
django_1     |   File "/usr/local/lib/python2.7/site-packages/gevent/builtins.py", line 93, in __import__
django_1     |     result = _import(*args, **kwargs)
django_1     |   File "/usr/src/app/publicmapping/wsgi.py", line 16, in <module>
django_1     |     application = get_wsgi_application()
django_1     |   File "/usr/local/lib/python2.7/site-packages/django/core/wsgi.py", line 13, in get_wsgi_application
django_1     |     django.setup(set_prefix=False)
django_1     |   File "/usr/local/lib/python2.7/site-packages/django/__init__.py", line 22, in setup
django_1     |     configure_logging(settings.LOGGING_CONFIG, settings.LOGGING)
django_1     |   File "/usr/local/lib/python2.7/site-packages/django/conf/__init__.py", line 56, in __getattr__
django_1     |     self._setup(name)
django_1     |   File "/usr/local/lib/python2.7/site-packages/django/conf/__init__.py", line 41, in _setup
django_1     |     self._wrapped = Settings(settings_module)
django_1     |   File "/usr/local/lib/python2.7/site-packages/django/conf/__init__.py", line 129, in __init__
django_1     |     raise ImproperlyConfigured("The SECRET_KEY setting must not be empty.")
django_1     | ImproperlyConfigured: The SECRET_KEY setting must not be empty.
django_1     | [2018-02-13 21:02:17 +0000] [8] [INFO] Worker exiting (pid: 8)
django_1     | [2018-02-13 21:02:17 +0000] [9] [INFO] Worker exiting (pid: 9)
django_1     | [2018-02-13 21:02:17 +0000] [1] [INFO] Shutting down: Master
django_1     | [2018-02-13 21:02:17 +0000] [1] [INFO] Reason: Worker failed to boot.
```

tnation14 commented on scripts/configure_va_demo:
```diff
@@ -0,0 +1,58 @@
+function recreate_database() {
+    echo "Drop district_builder database"
+    docker-compose \
+        exec -T postgres gosu postgres dropdb district_builder
```
Might be good to add an `--if-exists` flag to `dropdb` here.

tnation14 reviewed (approved):
All good pending the two changes I suggested

ddohler committed:
Generate config_settings.py in scripts/update

ddohler commented:
I'm going to go ahead and merge so that I can rebase #513 on this. The changes to `scripts/update` allowed Travis to make it through to running tests, so I think there's a reasonable chance that that PR will be able to pass after these changes are incorporated.
